src/main/java/com/google/netpcapanalysis/dao/FrequencyDaoImpl.java

```diff
@@ -0,0 +1,148 @@
+import com.google.netpcapanalysis.interfaces.dao.FrequencyDao;
+
+import java.nio.file.Path;
+import java.nio.file.Paths;
+
+public class FrequencyDaoImpl implements FrequencyDao {
+  private ArrayList<PCAPdata> allPCAP; 
+  private ArrayList<PCAPdata> finalFreq;
+  private HashMap<String, PCAPdata> finalMap;
+
+  private String myip = "";
+  private boolean first = true;
+  private int NUM_NODES = 15;
+
+  public FrequencyDaoImpl(ArrayList<PCAPdata> packets) {
+    allPCAP = packets; 
+  }
+
+  /* Retrieve all necessary entities, processes, and puts into final arraylist */
+  public void loadFrequency() {
+    findMyIP();
+    processData();
+    putFinalFreq();
+  }
+
+  public ArrayList<PCAPdata> getAllPCAP() {
+    return allPCAP;
+  }
+
+  /* Find local ip address based on highest recurring IP address */
+  private void findMyIP() {
+    HashMap<String, Integer> hm = new HashMap<String, Integer>();
+    for (PCAPdata packet : allPCAP) {
+      // source
+      if (hm.containsKey(packet.source)) { 
+        // if IP already exists, increment
+        hm.merge(packet.source, 1, Integer::sum);
+      }
+      else {
+        hm.put(packet.source, 1);
+      }
+      // destination
+      if (hm.containsKey(packet.destination)) { 
+        // if IP already exists, increment
+        hm.merge(packet.destination, 1, Integer::sum);
+      }
+      else {
+        hm.put(packet.destination, 1);
+      }
+    }
+    // find largest recurrence
+    myip = Collections.max(hm.entrySet(), Map.Entry.comparingByValue()).getKey();
+  }
+
+  public String getMyIP() {
+    return myip;
+  }
+
+  /* Fills out finalMap with frequencies based on IP address only (not protocol). */
+  private void processData(){
+    finalMap = new HashMap<String, PCAPdata>();
+    for (PCAPdata packet : allPCAP) {
+      String srcip = packet.source;
+      String dstip = packet.destination;
+      String protocol = "";
+      String outip = "";
+
+      if (srcip.equals(myip)) {
+        outip = dstip;
+      }
+      else {
+        outip = srcip;
+      }
+      
+      // PCAPdata takes in (source, destination, domain, location, protocol, size, flagged, frequency) 
+      if (finalMap.containsKey(outip)){
+        // retrieve current value with outip and increments frequency
+        PCAPdata currPCAP = finalMap.get(outip);
+        currPCAP.incrementFrequency();
+      }
+      else {
+        PCAPdata tempPCAP = new PCAPdata(myip, outip, "", "", packet.protocol, packet.size, packet.flagged, packet.frequency); 
+        finalMap.put(outip, tempPCAP);
+      }
+    }
+  }
+
+  public HashMap<String, PCAPdata> getFinalMap() {
+    return finalMap;
+  }
+
+  /* Transfers all unique connections to an arraylist for return. */
+  private void putFinalFreq() {
+    finalFreq = new ArrayList<PCAPdata>();
+    ArrayList<PCAPdata> allValues = new ArrayList<PCAPdata>();
+
+    for (PCAPdata packet : finalMap.values()) {
+      allValues.add(packet);
```

Comment:
you are only using the frequency in the packet object right, so just store the frequency in the map directly instead of this round about way.
`finalMap = new HashMap<String, Integer>();`